### DISCUSSION TITLE
nomadnet - ttyUSB0 - Attempting to use a port that is not open

### DISCUSSION THREAD

**Discussion by TheTimC** [markqvist/Reticulum#481] [Help & Questions]
Hi folks, I'm hoping you, or Mark, can help me. I am a newbie to all this, but my intention is to build anumber of rnodes for the local area to allow country folks to keep in contact during emergency situations. Initially, I just want to get these working so I can communicate on a closed LoRaWAN network.
I currently have three heltec LoRa 32 v2 boards, and I have successfully installed the rnode firmware on them. 
I am not a great Linux user, but I am persistent enough to get by. 

The issue:
When launching nomadnet, it is not connecting to the ttyUSB0 interface. At first I had access denied erros in the logs, but I resolved that by creating a dialup group, adding my user and then restarting.
Now the issue is that I get a "Attempting to use a port that is not open" error. I'll paste the log below.
All help gratefully received.
Tim.

> [2024-03-29 17:19:00] [Notice] Opening serial port /dev/ttyUSB0...
> [2024-03-29 17:19:02] [Error] Could not detect device for RNodeInterface[RNode LoRa Interface]
> [2024-03-29 17:19:02] [Error] A serial port error occurred, the contained exception was: [Errno 9] Bad file descriptor
> [2024-03-29 17:19:02] [Error] The interface RNodeInterface[RNode LoRa Interface] experienced an unrecoverable error and is now offline.
> [2024-03-29 17:19:02] [Error] Reticulum will attempt to reconnect the interface periodically.
> [2024-03-29 17:19:02] [Notice] Serial port /dev/ttyUSB0 is now open
> [2024-03-29 17:19:02] [Error] Could not open serial port for interface RNodeInterface[RNode LoRa Interface]
> [2024-03-29 17:19:02] [Error] The contained exception was: Attempting to use a port that is not open
> [2024-03-29 17:19:02] [Error] Reticulum will attempt to bring up this interface periodically

**Comment by faragher** [comment]
The dialout group is a big one. Good catch.

So, firmware is set up. Does it display the UI correctly? (I think Heltec boards have displays)

Have you set these up to use Bluetooth? If they're set to use Bluetooth they disable the USB serial connection.

The errors look like it's trying to open the port, not finding the device, and the error is really that it failed to initialize. 

 Is this the proper port? If it's /dev/ttyUSB0, then we need to keep digging, but if you're accidentally polling a mouse or the like and the RNode is actually on, say, /dev/ttyACM0, then it's an easy fix. I like easy fixes. :D

If you're not sure what it could be called, you can use ls `/dev/tty*` without the RNode connected, then with it connected. The difference is the device.

There are some other test we can do to make sure the RNode is functioning, but if you've taken care of dialout, then my next steps would be checking the port name and, it's happened to me too often, making sure you have a data cable for USB, not a charging cable.

**Reply by TheTimC** [reply]
Hi there. Thanks for your reply.
So, yes, I have confirmed it is definitely ttyUSB0. I had also done this by running sudo dmesg -w and then plugging it in. Also, this is the port that was reported when I flashed the rnode firmware onto the device. Also, this is the same USB cable I used when flashing it, so I know it is a data cable. Also, yes, the heltec has a display and I can see the RNODE firmware is up and running.
Also, I have successfully got it to connect on my phone with the Sideband app via a USB-C to USB-A adapter, although I did have to set the Radio options in the Hardware/RNode section, as the defaults were blank. Without another device to connect to, I haven't been able to test communicaiton, but the display does now change on the heltec to the Sideband logo, so I assume the phone one is working, I just need to get the Linux one working.
Interesting for your query about bluetooth. I didn't do anything with bluetooth at first, but I did try pairing it with bluetooth via the rnodeconf --bluetooth-pair /dev/ttyUSB0 command. It gives pairing codes, but then seems to fail to pair each time. Do I need to do something different to disable bluetooth on the heltec?
Grateful for any other suggestions, thanks.
Tim.

--EDIT
I just tried disabling bluetooth on the device using rnodeconf --bluetooth-off /dev/ttyUSB0. It said it had disabled bluetooth, but nomadnet still gives the same error.
Also, if it is useful, here is the output of rnodeconf --info /dev/ttyUSB0

> [21:01:33] Opening serial port /dev/ttyUSB0...
> [21:01:37] Device connected
> [21:01:37] Current firmware version: 1.70
> [21:01:37] Reading EEPROM...
> [21:01:37] EEPROM checksum correct
> [21:01:37] Device signature validated
> [21:01:37] 
> [21:01:37] Device info:
> [21:01:37]   Product            : Heltec LoRa32 v2 850 - 950 MHz (c0:c9:38)
> [21:01:37]   Device signature   : Validated - Local signature
> [21:01:37]   Firmware version   : 1.70
> [21:01:37]   Hardware revision  : 1
> [21:01:37]   Serial number      : 00:00:00:02
> [21:01:37]   Modem chip         : SX1276
> [21:01:37]   Frequency range    : 850.0 MHz - 950.0 MHz
> [21:01:37]   Max TX power       : 17 dBm
> [21:01:37]   Manufactured       : 2024-03-29 17:18:10
> [21:01:37]   Device mode        : Normal (host-controlled)

**Reply by faragher** [reply]
Right, I suppose you'd know the port from flashing the device. :)

Okay, if the Sideband logo appears, it's almost certainly working on that end. 

It's interesting that you can get information from the device but not connect to it using NomadNet. I'm going to assume that the device is functioning. I'm 90% sure this won't show us anything new, but have you tried `rnstatus`? It shouldn't work if NomadNet doesn't, but it might be easier to see error messages if any crop up.

It seems you're running NN stand-alone instead of running an rns deamon, correct?

Could you post your ~/.reticulum/config or at least the RNode Interface section?

I believe you've disabled Bluetooth.  I don't have a Heltec to test personally, but this sounds like it's on the computer side. Do you know what distro you're currently running?

**Reply by markqvist** [reply]
Also, without any other RNS apps or `rnsd` running, try to run `rnstatus -vvvvvvv`, which should give you full debug output for the startup and initialization of all interfaces.

**Reply by TheTimC** [reply]
Thanks both for your help.
To answer questions first, yes, I am running it standalone, I haven't tried running it as a service yet.
I've tried this on an Ubuntu Desktop USB boot, but I am now using my Linux Mint install on a Surface Pro 3.

OK, here is the RNode Interface section of the config file:
```
[[RNode LoRa Interface]]
    type = RNodeInterface
    interface_enabled = True
    port = /dev/ttyUSB0
    frequency = 867200000
    bandwidth = 125000
    txpower = 7
    spreadingfactor = 8
    codingrate = 5
```

And here is the output of rnstatus -vvvvvvv
```
[2024-03-29 22:25:22] [Debug] Started shared instance interface: Shared Instance[37428]
[2024-03-29 22:25:22] [Extra] Cleaning resource and packet caches...
[2024-03-29 22:25:22] [Verbose] Bringing up system interfaces...
[2024-03-29 22:25:22] [Extra] AutoInterface[Default Interface] No link-local IPv6 address configured for lo, skipping interface
[2024-03-29 22:25:22] [Extra] AutoInterface[Default Interface] Selecting link-local address fe80::25c:80c:440d:8431 for interface wlp1s0
[2024-03-29 22:25:22] [Extra] AutoInterface[Default Interface] Creating multicast discovery listener on wlp1s0 with address ff12:0:d70b:fb1c:16e4:5e39:485e:31e1
[2024-03-29 22:25:22] [Verbose] AutoInterface[Default Interface] discovering peers for 1.5 seconds...
[2024-03-29 22:25:23] [Notice] Opening serial port /dev/ttyUSB0...
[2024-03-29 22:25:26] [Error] Could not detect device for RNodeInterface[RNode LoRa Interface]
[2024-03-29 22:25:26] [Error] A serial port error occurred, the contained exception was: [Errno 9] Bad file descriptor
[2024-03-29 22:25:26] [Error] The interface RNodeInterface[RNode LoRa Interface] experienced an unrecoverable error and is now offline.
[2024-03-29 22:25:26] [Error] Reticulum will attempt to reconnect the interface periodically.
[2024-03-29 22:25:26] [Notice] Serial port /dev/ttyUSB0 is now open
[2024-03-29 22:25:26] [Verbose] Configuring RNode interface...
[2024-03-29 22:25:26] [Error] Could not open serial port for interface RNodeInterface[RNode LoRa Interface]
[2024-03-29 22:25:26] [Error] The contained exception was: Attempting to use a port that is not open
[2024-03-29 22:25:26] [Error] Reticulum will attempt to bring up this interface periodically
[2024-03-29 22:25:26] [Verbose] System interfaces are ready
[2024-03-29 22:25:26] [Verbose] Configuration loaded from /home/tim/.reticulum/config
[2024-03-29 22:25:26] [Verbose] Loaded 2 known destination from storage
[2024-03-29 22:25:26] [Verbose] Loaded Transport Identity from storage

 Shared Instance[37428]
    Status    : Up
    Serving   : 0 programs
    Rate      : 1.00 Gbps
    Traffic   : 0 B↑
                0 B↓

 AutoInterface[Default Interface]
    Status    : Up
    Mode      : Full
    Rate      : 10.00 Mbps
    Peers     : 0 reachable
    Traffic   : 0 B↑
                0 B↓

 RNodeInterface[RNode LoRa Interface]
    Status    : Down
    Mode      : Full
    Rate      : 0.00 bps
    Airtime   : 0.0% (15s), 0.0% (1h)
    Ch.Load   : 0.0% (15s), 0.0% (1h)
    Traffic   : 0 B↑
                0 B↓

[2024-03-29 22:25:26] [Debug] Saved packet hashlist in 0.16ms
[2024-03-29 22:25:26] [Debug] Saving path table to storage...
[2024-03-29 22:25:26] [Debug] Saved 0 path table entries in 0.11ms
[2024-03-29 22:25:26] [Debug] Saving tunnel table to storage...
[2024-03-29 22:25:26] [Debug] Saved 0 tunnel table entries in 0.09ms
[2024-03-29 22:25:26] [Debug] Saving 2 known destinations to storage...
[2024-03-29 22:25:26] [Debug] Saved known destinations to storage in 0.27ms
```

I hope this helps. Thanks,
Tim.

**Reply by markqvist** [reply]
Stupid question maybe, but is something connected to the Bluetooth port of the RNode *while* you try to connect over serial / USB as well? If that is the case, that wont work. When a host connects over Bluetooth, the serial / USB is temporarily disabled while the Bluetooth connection is active.

From the logs above, it looks like it *could* be the case, since the serial port itself can actually be opened, but then RNS gets no response back from the device (`[Error] Could not detect device for RNodeInterface[RNode LoRa Interface]`). So this might indicate that something else is connected to the device already.

**Reply by TheTimC** [reply]
Hi Mark,
I'm pretty sure that's not the case, but faragher had already suggested that above, so I used the command to turn Bluetooth-off. Also, the little Bluetooth icon is greyed out, except for the few times I did try to pair it. No joy.
I'm thinking it's maybe just something to do with the heltec board, I don't suppose you can get one for testing? 
Tim.

**Reply by TheTimC** [reply]
Mark, what is the best board I should get to test this using different hardware? Can you post a link to one on Amazon or ebay or something?

**Reply by markqvist** [reply]
I just remembered that I've seen this USB weirdness with Heltec boards on several occasions before. I had one board I ended up binning since it was just so intermittently weird I didn't want to deal with it. Another one works most of the time, as long as it has a good USB cable, but  seems ultra finicky with lower-quality cables. Then again, I have three others I use for testing the firmware that work completely fine all of the time, so really don't no what's up. Maybe just poor QC from the manufacturer.

The LilyGO LoRa32 version 2.1 boards are well supported, as are the T-Beam boards with both SX1276, SX1278, SX1262 and SX1268 chips, and they seem to have OK to decent quality assurance, from what I have seen. Boards using an ESP32-S3 microprocessor are not fully supported yet though, at least they are not easy to install.

**Reply by Dant3s** [reply]
I thought that the warning over Flashing Heltec Boards was about the USB issue.  I have 4 Boards, since is the only thing I could get my hands on in this country.

All present the same output, even if they work without issue. I even started to think about the quality of the USB cables since connecting with BT doesn't show any similar log output issue but the output is always there when using the USB.

**Reply by TheTimC** [reply]
Just a quick update. I ordered some more USB-C to USB-A adapters so I could use Sideband on multiple devices. I got this working and connecting fine, sending messages via the rnodes. I have also enabled bluetooth on one of them so I can just power it from a battery pack, that is also working fine.
So, my process seems to be fine as using my Linux machine to flash the firmware, then, once set up, I just use Sideband from mobile devices.
I guess I may need to revisit the Linux issue later, as I'll want to set up permanent nodes in people's houses at some point, but this will do for now.

One quick question, I seem to not be getting notifications on the mobile device when a new message comes in. I think I have it enabled, but nothing happens. Any ideas?

-- EDIT ---
Ah, I found that I get notifications once I've added the sender as a trusted sender. It would be good to get notifications of other things, like non-trusted messages and even if announces are heard?

**Reply by markqvist** [reply]
Thanks for the update @TheTimC! There's definitely room for improvements regarding the notification system. Maybe you could start a thread about it in the discussions section? In general I think it's nice to not spam the user with notifications by default, but being able to enable notifications for various events of interest would be a good addition.

**Reply by btalbot85** [reply]
So I am a beginner coder at best. But what I found is something to do with pyserial and flow control. Maybe these 2 links will shed light on it. I am only seeing the issue with the heltech v2 which stinks because I just bought a few of them. My ttgo works fine
https://github.com/pyserial/pyserial/issues/59
https://stackoverflow.com/questions/37288278/python-serial-attempting-to-use-a-port-that-is-not-open

**Reply by btalbot85** [reply]
@markqvist @TheTimC I just did further testing and can confirm this is a linux only issue. Same heltec v2 works fine with rnsd in windows. Seems to be a linux only issue and not device/firmware related.  The above link mentioned the same thing but its a bit beyond my skills to solve so far. Does appear Windows handles the control lines differently. Fingers crossed its a simple fix :) I really love my linux devices lol

**Reply by btalbot85** [reply]
Sorry for all the replies @TheTimC  @markqvist after more testing I have found if you set your heltec v2 rnode to tnc and make sure it exactly matches your rnode config in reticulum then start rnsd where youll see the port that is not open error looping you can then run rnodeconf /dev/ttyUSB0 -T it will fail after a few seconds  with some errors (sometimes i have to hit enter at the freq setting then run it the same command again but it does work lol. You will know it did when you see reconnected serial port.) Seems to be something with the mode switch and resetting the serial port to not open, not too sure. It does seem to be my timing when I run these and the port going back to available due to this. Not sure if its fixable. I also noticed if i time the reboot button just right before it says serial port opening it does the same thing

**Reply by TheTimC** [reply]
Oh, nice, thanks. Any chance you can share the commands to run to get it working?

**Reply by btalbot85** [reply]
So its a bit timing based on linux. If you setup in Windows it works out of the box.

First run             rnodeconf /dev/ttyUSB0 -T
Set everything it asks to match your reticulum config options for rnode. Youll then see your rnode switch into tnc mode.
run rnsd wait till you see the error start, it will keep looping which is what we want. Do not close
In another terminal run rnodeconf /dev/ttyUSB0 -T again and wait a few seconds for an error. Do not enter any settings, If after a few seconds you dont see an error or rnsd window does not say reconnected to serial press enter it will then error out. I seem to have best luck running this command right before you see the connecting to serial output in the rnsd window
do the rnodeconf /dev/ttyUSB0 -T again and do just like the last step all while keeping the one terminal with rnsd open and looking for the reconnect, it will connect after a few attempts
open nomadnet in another terminal (make sure you keep rnsd open in the other terminal) and enjoy

Its not pretty but it works lol. Due to it being about timing their really isn't a good way to script this.

If you need to get your rnode back into the normal mode for bluetooth or whatever else just do rnodeconf /dev/ttyUSB0 -N

**Comment by btalbot85** [comment]
I think I solved this. I noticed some weird issues with the OLED and decided to peek at the boards.h code. It shows rst pin as 23, but I found the rst pin is actually 14 by looking through pin definitions for heltec lora32 v2 on their GitHub. Swapped it out and now everything works in Linux as well. Just to confirm I went back to the original FW via the configurator and the error behavior came back. Pull request submitted with proper pin and evidence.

**Reply by PaulSchulz** [reply]
Is this fix in the released version of rns? 
-  rnsd 0.8.8
- rnodeconf 0.2.2
I am seeing this behavour, and the fix/manipulation also worked for me.
The 'renodeconf --autoinstall' confirms that the firmware version is 1.79

**Comment by rbomze** [comment]
I have the very same issue with a Heltec v2. Does not work on 2 linux computers, while a T114 v2 works fine.
Heltec v2 attaches at /dev/ttyUSB0, the T114 at /dev/ttyACM0.
Running at 'normal', host-controlled mode, not TNC.
- rnsd 0.8.9
- rnode 1.8.0

```
[2025-01-16 21:42:50] [Notice] Opening serial port /dev/ttyUSB0...
[2025-01-16 21:42:53] [Error] Could not detect device for RNodeInterface[RNode LoRa Interface]
[2025-01-16 21:42:53] [Notice] Serial port /dev/ttyUSB0 is now open
[2025-01-16 21:42:53] [Verbose] Configuring RNode interface...
[2025-01-16 21:42:53] [Error] Could not open serial port for interface RNodeInterface[RNode LoRa Interface]
[2025-01-16 21:42:53] [Error] The contained exception was: Attempting to use a port that is not open
```

**Reply by rbomze** [reply]
I was able to narrow down the problem a bit.
In `RNodeInterface.py`  `configure_device()` is run which runs `self.detect()`.
In the background a loop reading the serial interface expects some correct reply, which sets the `self.detected` flag to `True`. That does not happen here.

As a dirty hack i added a `self.detected = True` before the flag is checked and the serial device forcefully closed [before line 395]. Should be fixed properly.  

https://github.com/markqvist/Reticulum/blob/4fa335639caa1088b754d2b3413aae622172a709/RNS/Interfaces/RNodeInterface.py#L395

tl;dr The serial device works, but the detect function fails.

There seems to be a Heltec v2 and v2.1. Maybe the problem lies somewhere here. My device is v2.
Of course this is the problem seen from the rnsd side, the bug could be in Heltec v2's RNode code.
